File: app/src/main/java/mustafaozhan/github/com/mycurrencies/base/api/RxErrorHandlingCallAdapterFactory.java
package mustafaozhan.github.com.mycurrencies.base.api;

import java.io.IOException;
import java.lang.annotation.Annotation;
import java.lang.reflect.Type;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.Scheduler;
import io.reactivex.functions.Function;
import retrofit2.Call;
import retrofit2.CallAdapter;
import retrofit2.HttpException;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;

/**
 * Created by Mustafa Ozhan on 7/10/18 at 9:46 PM on Arch Linux wit Love <3.
 */
public class RxErrorHandlingCallAdapterFactory extends CallAdapter.Factory {
    private final RxJava2CallAdapterFactory original;

    private RxErrorHandlingCallAdapterFactory() {
        original = RxJava2CallAdapterFactory.create();
    }

    private RxErrorHandlingCallAdapterFactory(final Scheduler scheduler) {
        original = RxJava2CallAdapterFactory.createWithScheduler(scheduler);
    }

    public static CallAdapter.Factory create() {
        return new RxErrorHandlingCallAdapterFactory();
    }

    public static CallAdapter.Factory create(final Scheduler io) {
        return new RxErrorHandlingCallAdapterFactory(io);
    }

    @Override
    public CallAdapter<?, ?> get(Type returnType, Annotation[] annotations, Retrofit retrofit) {
        return new RxCallAdapterWrapper<>(retrofit, original.get(returnType, annotations, retrofit));
    }


    private static class RxCallAdapterWrapper<R> implements CallAdapter<R, Observable<R>> {
        private final Retrofit retrofit;
        private final CallAdapter<R, ?> wrapped;

        public RxCallAdapterWrapper(Retrofit retrofit, CallAdapter<R, ?> wrapped) {
            this.retrofit = retrofit;
            this.wrapped = wrapped;
        }

        @Override
        public Type responseType() {
            return wrapped.responseType();
        }

        @SuppressWarnings("unchecked")
        @Override
        public Observable<R> adapt(final Call<R> call) {
            return ((Observable) wrapped.adapt(call)).onErrorResumeNext(new Function<Throwable, ObservableSource>() {
                @Override
                public Observable apply(final Throwable throwable) {
                    return Observable.error(asRetrofitException(throwable));
                }
            });
        }

        private RetrofitException asRetrofitException(Throwable throwable) {
            // We had non-200 http error
            if (throwable instanceof HttpException) {
                HttpException httpException = (HttpException) throwable;
                Response response = httpException.response();
                return RetrofitException.httpError(response.raw().request().url().toString(), response, retrofit);
            }
            // A network error happened
            if (throwable instanceof IOException) {
                return RetrofitException.networkError((IOException) throwable);
            }

            // We don't know what happened. We need to simply convert to an unknown error

            return RetrofitException.unexpectedError(throwable);
        }
    }
}

File: app/src/main/java/mustafaozhan/github/com/mycurrencies/base/api/RxErrorHandlingCallAdapterFactory.kt
package mustafaozhan.github.com.mycurrencies.base.api

import android.annotation.SuppressLint
import io.reactivex.Observable
import io.reactivex.Scheduler
import retrofit2.Call
import retrofit2.CallAdapter
import retrofit2.HttpException
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import java.io.IOException
import java.lang.reflect.Type

/**
 * Created by Mustafa Ozhan on 7/10/18 at 9:46 PM on Arch Linux wit Love <3.
 */
class RxErrorHandlingCallAdapterFactory : CallAdapter.Factory {
    companion object {
        fun create(): CallAdapter.Factory = RxErrorHandlingCallAdapterFactory()
        fun create(io: Scheduler): CallAdapter.Factory = RxErrorHandlingCallAdapterFactory(io)
    }

    private val original: RxJava2CallAdapterFactory

    private constructor() {
        original = RxJava2CallAdapterFactory.create()
    }

    private constructor(scheduler: Scheduler) {
        original = RxJava2CallAdapterFactory.createWithScheduler(scheduler)
    }

    override fun get(returnType: Type, annotations: Array<Annotation>, retrofit: Retrofit): CallAdapter<*, *> {
        val wrapped = original.get(returnType, annotations, retrofit) as? CallAdapter<out Any, *>
        return RxCallAdapterWrapper(retrofit, wrapped)
    }

    private class RxCallAdapterWrapper<R>(
        private val retrofit: Retrofit,
        private val wrapped: CallAdapter<R, *>?
    ) : CallAdapter<R, Observable<R>> {

        override fun responseType(): Type? {
            return wrapped?.responseType()
        }

        @SuppressLint("CheckResult")
        @Suppress("UNCHECKED_CAST")
        override fun adapt(call: Call<R>): Observable<R>? {
            val adapted = (wrapped?.adapt(call) as? Observable<R>)
            adapted?.onErrorResumeNext { throwable: Throwable ->
                Observable.error(asRetrofitException(throwable))
            }

            return adapted
        }

        private fun asRetrofitException(throwable: Throwable) = when (throwable) {
            is HttpException -> RetrofitException.httpError(
                throwable.response().raw().request().url().toString(),
                throwable.response(),
                retrofit
            )
            is IOException -> RetrofitException.networkError(throwable)
            else -> RetrofitException.unexpectedError(throwable)
        }
    }
}
